cmd/mdatagen/validate.go

```diff
@@ -29,10 +30,20 @@ func (md *metadata) Validate() error {
 	return errs
 }
 
+// typeRegexp is used to validate the type of a component.
+// A type must start with an ASCII alphabetic character and
+// can only contain ASCII alphanumeric characters and '_'.
+// This must be kept in sync with the regex in component/config.go.
+var typeRegexp = regexp.MustCompile(`^[a-zA-Z][0-9a-zA-Z_]*$`)
+
 func (md *metadata) validateType() error {
```

Comment:
I would call `NewType` and ignore return value, only return error.

Reply:
Agree that that would be better, but it's not easy to do: it would mean adding a `replace` statement on `cmd/mdatagen/go.mod` for `component`, and that makes `go install` fail. See e.g. open-telemetry/opentelemetry-collector-contrib/issues/27855 for a recent instance of this issue with `telemetrygen`